wallet/src/wallet.rs

```diff
@@ -0,0 +1,165 @@
+        Self::load_wallet(db)
+    }
+
+    fn load_wallet(db: WalletStorageImpl) -> Result<Self, WalletError> {
+        let txs = db.read_transactions().map_err(WalletError::DatabaseError)?;
+        let utxo = db.read_utxo_set().map_err(WalletError::DatabaseError)?;
+
+        Ok(Wallet { db, txs, utxo })
+    }
+
+    pub fn get_database(&self) -> &WalletStorageImpl {
+        &self.db
+    }
+
+    #[allow(dead_code)] // TODO remove
+    fn add_transaction(&mut self, tx: Transaction, state: TxState) -> Result<(), WalletError> {
+        let tx_id = tx.get_id();
+
+        if self.txs.contains_key(&tx_id) {
+            return Err(WalletError::DuplicateTransaction(tx_id));
+        }
+
+        let mut db_tx = self.db.transaction_rw(None).map_err(WalletError::DatabaseError)?;
+
+        let mut wallet_tx = WalletTx::new(tx, state);
+        wallet_tx.set_order(self.txs.len() as i64);
+
+        db_tx.set_transaction(&tx_id, &wallet_tx).map_err(WalletError::DatabaseError)?;
+        db_tx.commit().map_err(WalletError::DatabaseError)?;
+
+        self.txs.insert(tx_id, wallet_tx);
+
+        Ok(())
+    }
+
+    // TODO fix incompatibility between borrowing mut self and the database transaction
+    #[allow(dead_code)] // TODO remove
+    fn add_to_utxos(
+        &mut self,
+        tx: &Transaction,
+        db_tx: &mut WalletStorageTxRwImpl,
+    ) -> Result<(), WalletError> {
+        for (i, output) in tx.outputs().iter().enumerate() {
+            // Check if this output belongs to this wallet or it is watched
+            if self.is_available_for_spending(output) && self.is_mine_or_watched(output) {
+                let outpoint = OutPoint::new(tx.get_id().into(), i as u32);
+                let utxo = Utxo::new(output.clone(), false, utxo::UtxoSource::Mempool);
+                self.utxo.insert(outpoint.clone(), utxo.clone());
+                db_tx.set_utxo(&outpoint, utxo).map_err(WalletError::DatabaseError)?;
+            }
+        }
+        Ok(())
+    }
+
+    #[allow(dead_code)] // TODO remove
+    fn is_available_for_spending(&self, _txo: &TxOutput) -> bool {
+        // TODO implement
+        true
+    }
+
+    #[allow(dead_code)] // TODO remove
+    fn is_mine_or_watched(&self, _txo: &TxOutput) -> bool {
+        // TODO implement
+        true
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use common::chain::{GenBlock, Transaction};
+    use common::primitives::H256;
+    use std::time::SystemTime;
+
+    #[test]
+    fn in_memory_wallet() {
+        let wallet = Wallet::open_wallet_in_memory();
+        assert!(wallet.is_ok())
+    }
+
+    #[test]
+    fn wallet_transactions() {
+        let mut wallet_path = std::env::temp_dir();
+        let current_time =
+            SystemTime::now().duration_since(SystemTime::UNIX_EPOCH).unwrap().as_millis();
+        wallet_path.push(format!("wallet_transactions_{current_time}.sqlite"));
```

Comment:
While this might work, it is an invitation to future problems where someone copies this test and forgets to change the string, and then we have unexpected errors because of a 1/1000 chance both tests run in parallel and click at the same time. 

There are already solutions for this. Please see how `TempDir` is used for lmdb and config files. Grep the code. It's much simpler to use than this overly complex solution, in comparison.

Reply:
Yeah, it's better with the `TempDir`